### PR TITLE
Refactoring 5/N: Promise

### PR DESCRIPTION
**Pull request by mcheshkov**

### CODE REVIEW

File: lib/cluster_process.js [hunk 1]
```diff
@@ -218,21 +214,18 @@ class ClusterProcess extends EventEmitterEx {
...
+     * Resolves when ClusterProcess done initialization.
      * @this {ClusterProcess}
-     * @param {Function} fn
-     * @returns {Function}
+     * @returns {Promise<void>}
      */
-    static whenInitialized(fn){
-        return /** @this {ClusterProcess} */function() {
+    whenInitialized(){
+        return new Promise(resolve => {
```
Comment (SlNPacifist) [inline lib/cluster_process.js]: It'd be nice to have single promise here and not create it for every call. Something like
```
get whenInitialized() {
    if (!this._initPromise) {
        this._initPromise = new Promise(resolve => this.once('initialized', resolve));
    }
    return this._initPromise;
}
```

File: lib/master.js [hunk 2]
```diff
@@ -177,14 +177,11 @@ class Master extends ClusterProcess {
...
-     *          // fn
-     *      }, master);
+     *      master.getWorkersArray().forEach(fn);
      */
     forEach(fn) {
-        this.getWorkersArray().forEach(function(worker) {
-            fn.call(this, worker);
-        }, this);
+        this.getWorkersArray()
+            .forEach(fn);
```
Comment (SlNPacifist) [inline lib/master.js]: This changes previous behaviour and requires minor version bump at least.

File: lib/worker.js [hunk 3]
```diff
@@ -90,19 +88,14 @@ class Worker extends ClusterProcess {
         this.emit('foreign properties received');
     }
 
-    /**
-     * @param {Function} fn
-     */
-    static whenForeignPropertiesReceived(fn) {
-        return /** @this {Worker} */function() {
+    whenForeignPropertiesReceived() {
```
Comment (SlNPacifist) [inline lib/worker.js]: Same comment as for `whenInitialized`

File: lib/master.js [hunk 4]
```diff
@@ -243,45 +240,40 @@ class Master extends ClusterProcess {
...
+     * @returns {Promise<void>}
      */
-    waitForWorkers(wids, event, callback) {
-        const self = this;
+    waitForWorkers(wids, event) {
         const pendingWids = new Set(wids);
 
-        function onWorkerState(worker) {
-            const wid = worker.wid;
-            if (pendingWids.has(wid)) {
```
Comment (SlNPacifist) [inline lib/master.js]: BTW: we do not need this check with Set.

File: lib/master.js (hunk 2 above)
Comment (mcheshkov) [reply lib/master.js]: I'm planning to release this with major version bump and worker pools. Anyway, I've bumped minimal node version.

**Pull request merged**